source/particles/particle_handler.cc

```diff
@@ -0,0 +1,1152 @@
+#endif
+
+    sorted_particles_map.insert(sorted_particles.begin(),sorted_particles.end());
+
+    for (unsigned int i=0; i<particles_out_of_cell.size(); ++i)
+      remove_particle(particles_out_of_cell[i]);
+
+    particles.insert(sorted_particles_map.begin(),sorted_particles_map.end());
+    update_cached_numbers();
+  }
+
+
+
+#ifdef DEAL_II_WITH_MPI
+  template <int dim, int spacedim>
+  void
+  ParticleHandler<dim,spacedim>::exchange_ghost_particles()
+  {
+    // Nothing to do in serial computations
+    if (dealii::Utilities::MPI::n_mpi_processes(triangulation->get_communicator()) == 1)
+      return;
+
+    // First clear the current ghost_particle information
+    ghost_particles.clear();
+
+    std::map<types::subdomain_id, std::vector<particle_iterator> > ghost_particles_by_domain;
+
+    const std::set<types::subdomain_id> ghost_owners = triangulation->ghost_owners();
+    for (auto ghost_domain_id = ghost_owners.begin(); ghost_domain_id != ghost_owners.end(); ++ghost_domain_id)
+      ghost_particles_by_domain[*ghost_domain_id].reserve(static_cast<typename std::vector<particle_iterator>::size_type> (particles.size()*0.25));
+
+    std::vector<std::set<unsigned int> > vertex_to_neighbor_subdomain(triangulation->n_vertices());
+
+    typename Triangulation<dim,spacedim>::active_cell_iterator
+    cell = triangulation->begin_active(),
+    endc = triangulation->end();
+    for (; cell != endc; ++cell)
+      {
+        if (cell->is_ghost())
+          for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+            vertex_to_neighbor_subdomain[cell->vertex_index(v)].insert(cell->subdomain_id());
+      }
+
+    cell = triangulation->begin_active();
+    for (; cell != endc; ++cell)
+      {
+        if (!cell->is_ghost())
+          {
+            std::set<unsigned int> cell_to_neighbor_subdomain;
+            for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+              {
+                cell_to_neighbor_subdomain.insert(vertex_to_neighbor_subdomain[cell->vertex_index(v)].begin(),
+                                                  vertex_to_neighbor_subdomain[cell->vertex_index(v)].end());
+              }
+
+            if (cell_to_neighbor_subdomain.size() > 0)
+              {
+                const particle_iterator_range particle_range = particles_in_cell(cell);
+
+                for (std::set<types::subdomain_id>::const_iterator domain=cell_to_neighbor_subdomain.begin();
+                     domain != cell_to_neighbor_subdomain.end(); ++domain)
+                  {
+                    for (typename particle_iterator_range::iterator particle = particle_range.begin(); particle != particle_range.end(); ++particle)
+                      ghost_particles_by_domain[*domain].push_back(particle);
+                  }
+              }
+          }
+      }
+
+    send_recv_particles(ghost_particles_by_domain,
+                        ghost_particles);
+  }
```

Comment:
and you need the `#endif` here, and another `#ifdef` in front of the next function, since that is definitely commented out. (Without harm in the latter case since it's a private function of the class. So it can't be called from anywhere else, and it's not called from within this class if we're not using MPI.)